libraries/chain/include/eosio/chain/block.hpp

```diff
@@ -103,6 +103,79 @@ namespace eosio { namespace chain {
    };
    using signed_block_ptr = std::shared_ptr<signed_block>;
 
+   struct pruned_transaction_receipt : public transaction_receipt_header {
+
+      pruned_transaction_receipt():transaction_receipt_header(){}
+      pruned_transaction_receipt(const transaction_receipt&, bool legacy);
+      explicit pruned_transaction_receipt( const transaction_id_type& tid ):transaction_receipt_header(executed),trx(tid){}
+      explicit pruned_transaction_receipt( const pruned_transaction& ptrx ):transaction_receipt_header(executed),trx(ptrx){}
+
+      fc::static_variant<transaction_id_type, pruned_transaction> trx;
+
+      std::size_t maximum_pruned_pack_size( pruned_transaction::compression_type segment_compression ) const;
+
+      digest_type digest()const {
+         digest_type::encoder enc;
+         fc::raw::pack( enc, status );
+         fc::raw::pack( enc, cpu_usage_us );
+         fc::raw::pack( enc, net_usage_words );
+         if( trx.contains<transaction_id_type>() )
+            fc::raw::pack( enc, trx.get<transaction_id_type>() );
+         else
+            fc::raw::pack( enc, trx.get<pruned_transaction>().packed_digest() );
+         return enc.result();
+      }
+   };
+
+   struct pruned_block : public signed_block_header{
+   private:
+      pruned_block( const pruned_block& ) = default;
+   public:
+      enum class prune_state_type : uint8_t { incomplete, complete, complete_legacy };
+
+      pruned_block() = default;
+      explicit pruned_block( const signed_block_header& h ):signed_block_header(h){}
+      pruned_block( const signed_block&, bool legacy );
+      pruned_block( pruned_block&& ) = default;
+      pruned_block& operator=(const pruned_block&) = delete;
+      pruned_block clone() const { return *this; }
+
+      fc::enum_type<uint8_t,prune_state_type> prune_state{prune_state_type::complete_legacy};
+      deque<pruned_transaction_receipt>       transactions; /// new or generated transactions
+      extensions_type                         block_extensions;
+
+      std::size_t maximum_pruned_pack_size( pruned_transaction::compression_type segment_compression ) const;
+
+      template<typename Stream>
+      void pack_with_padding(Stream& stream, pruned_transaction::compression_type segment_compression) {
+         std::size_t padded_size = maximum_pruned_pack_size( segment_compression );
+         pack_with_padding(stream, segment_compression, padded_size);
+      }
+      template<typename Stream>
+      void pack_with_padding(Stream& stream, pruned_transaction::compression_type segment_compression, std::size_t padded_size) {
+         fc::raw::pack(stream, fc::unsigned_int(padded_size));
+         std::size_t pos = stream.tellp();
+         fc::raw::pack(stream, *this);
+         std::size_t actual_size = stream.tellp() - pos;
+         for(std::size_t i = 0; i < padded_size - actual_size; ++i) {
```

Comment:
How about:
```cpp
for(std::size_t i = actual_size; i < padded_size; ++i ) {
```
In case the caller provides a `padded_size` that is too small?

Reply:
It's an error if the padded size is too small.  It means that you've just overwritten data that you shouldn't.